src/miner.cpp

```diff
@@ -814,7 +814,7 @@ void BlockAssembler::addPackageTxs(int &nPackagesSelected, int &nDescendantsUpda
     }
 }
 
-void IncrementExtraNonce(CBlock* pblock, const CBlockIndex* pindexPrev, unsigned int& nExtraNonce)
+void IncrementExtraNonce(CBlock* pblock, unsigned int nHeight, unsigned int& nExtraNonce)
 {
     // Update nExtraNonce
     static uint256 hashPrevBlock;
```

Comment:
As I hinted at (although this probably is going to show up first in the review comments), this I believe is where the cs_main is protecting.  It doesn't necessarily need to be cs_main and we could probably use a different lock to isolate to just this static, but here's the issue.

`IncrementExtraNonce` can be called from multiple mining threads, yet there's a static variable in there to store the latest hashPrevBlock and reset nExtraNonce if the block has changed.  So that introduces a race condition if two threads are coming in here at the same time, as one may change that static when the previous thread is looking at it.

Should this be locked by cs_main?  Probably not.  Should IncrementExtraNonce be locked?  Yes.  (Statics are the #1 warning sign of a need to lock)

Reply:
Ah, I hadn't thought of the static.
https://github.com/Veil-Project/veil/blob/e8c4d290c647b8e3363de49c440587fe091e963b/src/miner.cpp#L815-L820
It looks like this is the entire usage. This could be replaced with a `std::atomic<T>::exchange`, with `nExtraNonce` being reset if the previous value was not `pblock->hashPrevBlock`. Does that seem sufficient in place of a full lock, if there's nothing else?

Reply:
Let me ramp myself up on std::atomic(), but I'll definitely consider it [The bulk of my multi-threading development came with C, and I never really re-learned solutions with C++]

Reply:
Moved to https://github.com/Veil-Project/veil/pull/921